aeppl/math_stat.py

```diff
@@ -0,0 +1,62 @@
+import aesara
+import aesara.tensor as at
+from aesara.graph.basic import ancestors
+from aesara.graph.rewriting.basic import EquilibriumGraphRewriter, node_rewriter
+from aesara.scalar.basic import Add
+from aesara.tensor.elemwise import Elemwise
+from aesara.tensor.random.basic import NormalRV, normal
+from aesara.tensor.var import TensorConstant
+
+from aeppl.rewriting import logprob_rewrites_db
+
+
+@node_rewriter((Elemwise,))
+def add_independent_normals(fgraph, node):
+    if not isinstance(node.op.scalar_op, Add):
+        return None
+
+    X_rv, Y_rv = node.inputs
+
+    if not (X_rv.owner and Y_rv.owner) or not (
+        isinstance(X_rv.owner.op, NormalRV) and isinstance(Y_rv.owner.op, NormalRV)
+    ):
+        return None
+
+    # check independence
+    intersection = list(set(ancestors([X_rv])).intersection(set(ancestors([Y_rv]))))
+    intersection = [
+        node for node in intersection if not isinstance(node, TensorConstant)
+    ]
+
+    if intersection != []:
+        return None
+
+    old_rv = node.outputs[0]
+
+    mu_x, sigma_x = X_rv.owner.inputs[-2:]
+    mu_y, sigma_y = Y_rv.owner.inputs[-2:]
+
+    new_node = normal.make_node(
+        X_rv.owner.inputs[0],  # temporary rng?
```

Comment:
This is a tough one we've been trying to deal with more generally.  The real complications come from any "updates" that are associated with the RNG objects.

More specifically, RNG objects are usually `SharedVariable`s, and those can have `SharedVariable.default_update` attributes that hold onto other `Variable`s (i.e. graphs representing the new value of the `SharedVariable` after each call to a compiled Aesara function with this update).  In the case of RNG objects created by `RandomStream`s, those default updates are the RNG objects output after drawing a sample from a `RandomVariable` node.  In other words, the updates mechanism replaces a shared RNG objects with a copy of the RNG object after a sample has been drawn using it, so the Aesara updates mechanism is emulating an in-place update of the RNG (e.g. just as `rng.normal()` automatically updates the internal state of `rng` in NumPy).

When performing replacements like this, it's possible that `rng = X_rv.owner.inputs[0]` will have a `rng.default_update` containing the original `X_rv` graph, and, if someone attempted to `aesara.compile` the resulting `new_node`, Aesara would pick up that old graph from the re-used RNG's `default_update` attribute and add it to the compiled results.  We definitely don't want to have to sample some unrelated graphs just to update the RNG objects, especially when those update graphs aren't consistent with the underlying sampling process.

Anyway, I'm pointing this out because it's a general design issue and usability complication of which I'm trying to make more people aware&mdash;mostly so we can fix it/improve the usability.

In this exact case, we can probably just clone the RNG `SharedVariable`s and add our own `default_update`s to those (e.g. similar to how `RandomStream.gen` does).  Also, the graphs produced here are only supposed to be used as an intermediate representation for obtaining log-probabilities, so, as long as we don't expect people to actually compile and sample these graphs, the default updates shouldn't matter, and we can probably just clone the RNGs and remove their default updates altogether.

Reply:
Hmm, there's a lot of info here and I have the general picture. Thanks for the explanation.

When you say clone the RNG `SharedVariable`, do you mean:
```python
rng = X_rv.owner.inputs[0].clone()
rng.default_update = something
```
I'm confused because it feels like there are two RNGs here, one from `X_rv` and one from `Y_rv`... Or should I create a new RNG object akin to what's being done in `RandomStream.gen`?

Reply:
> When you say clone the RNG `SharedVariable`, do you mean:

Yeah, with one important difference, though:
```python
new_rng = X_rv.owner.inputs[0].clone()
# something more needs to be done to `new_rng...

new_node = normal.make_node(rng, *other_inputs)
new_rng.default_update = new_node.outputs[0]
```

In other words, it's the `new_rng` that needs to be updated with the values of the RNGs output by `new_node`.

> I'm also confused because it feels like there are two RNGs here, one from `X_rv` and one from `Y_rv`

Which RNG we use is really just a choice.  The only thing we really need to consider is the user-level seeding, and, as long as we choose an RNG from one of the existing `RandomVariable`s, we should maintain some consistency with the seeding.  If we use `RandomStream.gen`, then we can't generate a connection with the user's seeding unless we have their `RandomStream` instance.

Reply:
Also, because this whole `RandomVariable` + `SharedVariable.default_update` situation is such a mess, I've created this: https://github.com/aesara-devs/aesara/issues/1478.